_posts/2023-11-03-visual-testing-using-playwright.md

```diff
@@ -0,0 +1,201 @@
+  description="Manual Visual Identification of the elements"
+%}
+
+**Challenges of manual Visual Testing:-**
+Imagine you need to test your application on
+
+- 5 operating systems: windows, MacOS, Android, iOS, and Chrome
+- 5 popular browsers: chrome, Firefox, Internet Explorer (Windows only), Microsoft Edge (Windows Only), and Safari (Mac only)
+- 2 screen orientations for mobile devices: portrait and landscape
+- 10 standard mobile device display resolutions and 18 standard desktop/laptop display resolutions from XGA to 4G
+
+**1 Screen to test = 21 x (20+18) = 21 x 38 = 798 Unique Screen Configurations**
+
+Let’s say if your application has 100 pages to test
+
+**798 Screen Configurations x 100 Screens in-app = 79,800 Screen Configurations to test**
+
+Wouldn’t it be great if there was a way to automate this crazy-tedious process?
+
+Well, yes there is… a process is in place to do the tedious 
+
+
+Here we are with the problem statement from Visual Testing
+
+**🚨The problem** - Visual bugs are blocking revenue and resulting in not retaining customers.
+
+**🎯The goal** - Visual testing captures defects that other types of UI tests cannot [Functional, Manual & Regression].
+
+**💡The solution** - **Automated Visual Testing is a solution** and **Playwright** is one of the good option used in Backbase [**Applitools** or **Percy** are the other automated tools are available for visual testing]
+<br>
+<br>
+
+## **Automated Visual Testing using Playwright**
+
+To address the preceding challenges automated visual testing need to be implemented in place of ‘function testing’/’manual visual testing’ to **achieve efficient and reliable results in visual testing**
+
+- **Playwright** is an open-source test automation framework that enables end-to-end testing for modern web-apps
+- It uses languages - JavaScript, TypeScript, Python, Java, and C#
+- It also supports Chromium, Firefox, and WebKit-based browsers and can also run tests against an emulated mobile browser
+- OS support - Windows, Linux, and macOS
+
+
+**Snapshot Comparison Workflow:-**
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing6.png"
+%}
+
+**Simple Visual Snapshot Comparison Tests in Playwright**
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing7.png"
+%}
+
+When the **toHaveScreenshot()** function executes, it performs the following
+
+
+- It verifies if there is an existing screenshot file. If not present, it fails
+- It takes a new screenshot
+- It computes the difference between the two screenshots using the pixel match library
+- If there is any difference, take another screenshot
+- It computes the difference again
+- It writes an expected, actual, and diff file to the test-results folder
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing8.png"
+%}
+
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing9.png"
+%}
+
```

Comment:
Optional: the post ends a bit abruptly as I would expect some conclusion or key take-aways in the end. If possible, I would consider adding something after the last screens

Reply:
I have added conclusion/take-away points, can you please review